text/0069-api-v2.md

```diff
@@ -0,0 +1,171 @@
+
+### Client Version
+
+TiKV clients will have a new configuration choosing `API V1` or `API V2` mode on connecting to TiKV.
+
+TiKV will reject the request if the client mode is mismatched to the TiKV configuration. There is one exception: With TiKV in `API V2` mode, you can use `TxnClient` in `API V1` mode to access the TiDB data (key starting with `m` or `t`) which is invisible to the client in `API V2` mode.
+
+Note that in `API V2` mode, you can't specify CF in `RawClient`.
+
+Using TxnKV along with RawKV is safe in `API V2` mode.
+
+Listed below is the detailed compatibility matrix:
+
+|              | V1 Server | V2 Server |
+| ------------ | --------- | --------- |
+| V1 RawClient | Raw Data  | Forbidden |
+| V1 TxnClient | Txn Data  | TiDB Data |
+| V2 RawClient | Forbidden | Raw Data  |
+| V2 TxnClient | Forbidden | Txn Data  |
+
+### Keyspace
+
+When API V2 is enabled for the client, the user will be able to specify a keyspace for a session of `RawClient` or `TxnCient`. A keyspace is identified by a string, whose default value is 'default'.
+
+Keys in other keyspace will be invisible to the user.
+
+## Reference-level design
+
+### Key Encode
+
+Once the `API V2` is enabled, the key will be starting with either:
+
+1. `m` or `t`: TxnKV key. Used by TiDB.
+2. `x{keyspace prefix id}`: TxnKV key.
+3. `r{keyspace prefix id}`: RawKV key.
+4. `a{keyspace prefix id}`: RawKV atomic key, on which you can use `CompareAndSwap`.
+
+The `{keyspace prefix id}` is the [keyspace](https://github.com/tikv/rfcs/pull/39) prefix for seperating keys of different keyspace. It should be an vary-length integer whose highest bit of every byte denotes whether the next byte is still part of the integer. The client will fetch the prefix from PD by the keyspace name specified by the user when initializing the client, so it means that the keyspace prefix is valid during the session, in other words, the change on keyspace on PD will not take effect on running seesions.
+
+Note that in TxnKV, the key will be encoded by `Memory Comparable Encoding`. But since the `Memory Comparable Encoding` will not change the starting bytes and only add paddings, there won't be any overlap between RawKV and TxnKV.
+
+### RawKV Value Encode
+
+If the key is started with `r` or `a`, then the value can be either:
+
+1. `{0x0}{data}`
+2. `{0x1}{TTL expire timestamp}{data}`
+
+### PD
+
+Add a new http interface for adding, renaming, deleteing and querying the mapping from keyspace name to prefix id:
+
+```javascript
+// list all keyspaces
+GET /keyspaces
+[
+  {
+    name: "default",
+    id: "0"
```

Comment:
+1, an optional field to describe the keyspace is useful. Lots of things can be added upon keyspace, for example:

- users and their access privileges
- placement policies to physically isolate logically different keyspaces

we can do it later since these functionalities don't break the underlying key-value codec, they only affect the keyspace metadata.

Reply:
foundationdb and Cassandra don't provide such a field for keyspace:
- https://apple.github.io/foundationdb/developer-guide.html#developer-guide-sub-keyspaces
- https://docs.datastax.com/en/cql-oss/3.3/cql/cql_reference/cqlCreateKeyspace.html

In addition, we should also provide an API for modifying the description, which adds more complexity.

Reply:
Note that a FDB subspace is client-side only: there is no central registry. It is likely that some FDB users implement create a separate central registry (maybe just a document) that has a description.

I can see how without any real [auth system](https://github.com/tikv/rfcs/pull/40) in place there will always be a desire to mutate data because it always brings up questions relating to permission which are impossible to address.